File: resources/android/src/interaction/NativePHPChartsRadialSelection.kt
package com.donmanueldev.plugins.nativephp_charts.ui

import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import org.json.JSONObject

/** Serializes a confirmed pie/donut segment selection for NativePHP. */
internal object NativePHPChartsRadialSelection {
    /**
     * Emits once per caller-confirmed tap or accessibility action when `_select`
     * is bound. A radial segment maps to both series and point fields so it stays
     * compatible with the common version-1 selection payload.
     */
    fun dispatch(
        node: NativeUINode,
        configuration: NativePHPChartsRadialConfiguration,
        formatting: NativePHPChartsRadialFormatting,
        segment: NativePHPChartsRadialSegment,
    ) {
        if (configuration.onSelect == 0) return
        val chartType = configuration.kind.name.lowercase()
        val payload = JSONObject()
            .put("version", 1)
            .put("chart_type", chartType)
            .put("series_id", segment.id)
            .put("series_name", segment.label)
            .put("point_id", segment.id)
            .put("point_index", segment.index)
            .put("x_type", "category")
            .put("x", segment.label)
            .put("label", segment.label)
            .put("value", segment.value)
            .put("localized_value", formatting.value(segment.value))
            .toString()
        NativeUIBridge.sendTextChangeEvent(configuration.onSelect, node.id, payload)
    }
}

File: resources/ios/Sources/Interaction/NativePHPChartsRadialSelection.swift
import Foundation

/// Adapts radial identity to the shared version-one point-selection payload.
enum NativePHPChartsRadialSelection {
    static func payload(
        kind: NativePHPChartsRadialKind,
        segment: NativePHPChartsRadialSegment,
        formatter: NativePHPChartsRadialFormatter
    ) -> String? {
        NativePHPChartsSelectionPayload(
            chartType: kind.rawValue,
            seriesID: segment.id,
            seriesName: segment.label,
            pointID: segment.id,
            pointIndex: segment.index,
            xType: NativePHPChartsXAxisType.category.rawValue,
            x: .string(segment.label),
            label: segment.label,
            value: segment.value,
            localizedValue: formatter.value(segment.value)
        ).json()
    }
}

enum NativePHPChartsRadialAccessibility {
    static func summary(
        data: NativePHPChartsRadialDataSet,
        formatter: NativePHPChartsRadialFormatter,
        selectedSegment: NativePHPChartsRadialSegment?
    ) -> String {
        let visibleSegments = data.segments.prefix(12).map { segment in
            "\(segment.label): \(formatter.value(segment.value))"
        }
        let remainder = max(0, data.segments.count - visibleSegments.count)
        var summary = visibleSegments.joined(separator: ", ")

        if remainder > 0 {
            summary += ". (+\(remainder))"
        }
        if let selectedSegment {
            summary += ". \(selectedSegment.label), \(formatter.value(selectedSegment.value))"
        }

        return summary
    }
}
